### PR TITLE
Feature/mobile offline cache

### LINKED ISSUE #23: offline support
Application should support offline mode . i.e. create product ,client in offline mode.

### PR DESCRIPTION
**Pull request by souramoo**
This pull requests adds:

1. Caching functionality to allow display of invoice lists/etc when the mobile app goes offline (view-only at present).
2. A banner indicating offline status to the user.

This allows the mobile app to at least be usable when in an area with spotty internet connection, and provides a basis to improve on to fix #23

### CODE REVIEW

**Comment by hillelcoren** [comment]
We appreciate the PR however I'm not sure about adding a dependency for Hive. 

If you decide to correct it please also create the PR against the develop branch.

**Comment by hillelcoren** [comment]
Just to be clear.. Hive shouldn't be needed because the app is already caching all of the data. If we cache all response results it would cause the data to be cached twice.

**Comment by souramoo** [comment]
Hi,

Thanks - I've had a look around to see how I can remove hive by taking advantage of the app's existing cache. Unfortunately, as far as I can tell, the app does not appear to be already caching all of the data.

I came to this conclusion by first looking at the contents of the app's data folder:

```
1|redfin:/data/data/com.invoiceninja.app # ls app_flutter/                                                                                                                                                      
flutter_assets  invoiceninja__auth_state.json  invoiceninja__static_state.json  invoiceninja__ui_state.json  res_timestamp-160-1721032225752
```

None of these files contain any invoice data - although it does contain data for the auth, static state and UI in the respective files. I did a quick grep for one of my invoicee's names and it came up blank.

I also tried by replacing L#149 in lib/data/repositories/auth_repository.dart with a null response, after an app restart, on the invoice screen the app retains the "0 invoices" text despite having previously sync'd...

Please can you point me in the direction of where the app is caching the responses and I can take advantage of that?

**Comment by hillelcoren** [comment]
Please check that the cache option is enabled on Settings > Device Settings

**Comment by souramoo** [comment]
Thanks! Have removed hive completely now - and added a little green 2 second back online banner

File: pubspec.lock [hunk 1]
```diff
@@ -1335,18 +1367,18 @@ packages:
     dependency: transitive
     description:
       name: sentry
-      sha256: "19a267774906ca3a3c4677fc7e9582ea9da79ae9a28f84bbe4885dac2c269b70"
+      sha256: "57514bc72d441ffdc463f498d6886aa586a2494fa467a1eb9d649c28010d7ee3"
```
Comment (hillelcoren) [inline pubspec.lock:1370]: Why is the Sentry package changed?

File: lib/data/web_client.dart [hunk 2]
```diff
@@ -46,24 +54,29 @@ class WebClient {
...
     print('GET: $url');
 
-    final client = http.Client();
-    final http.Response response = await client.get(
-      Uri.parse(url),
-      headers: _getHeaders(url, token),
-    );
-    client.close();
+    final isOnline = await _isOnline();
+    print("isOnline: $isOnline");
```
Comment (hillelcoren) [inline lib/data/web_client.dart:58]: Please remove print statements

File: pubspec.lock (hunk 1 above)
Comment (souramoo) [reply pubspec.lock:1370]: I think my flutter dependency resolver updated version: "7.20.0" to version: "7.20.2"

**Comment by hillelcoren** [comment]
Thank you!

**Comment by hillelcoren** [comment]
This feature has been included with the latest version of the app but a number of people have reported seeing the "No connection" message even though they're online. Are you able to look into it?

**Comment by souramoo** [comment]
Sure! Will investigate further.

From: Hillel Coren ***@***.***>
Date: Saturday, 7 September 2024 at 22:23
To: invoiceninja/admin-portal ***@***.***>
Cc: Souradip Mookerjee ***@***.***>, Author ***@***.***>
Subject: Re: [invoiceninja/admin-portal] Feature/mobile offline cache (PR #666)

This feature has been included with the latest version of the app but a number of people have reported seeing the "No connection" message even though they're online. Are you able to look into it?

**Pull request merged**